### PR TITLE
[ROAD-856] fix: threading warnings

### PR DESCRIPTION
**Pull request by asaf92**
### Description

Fixed threading warnings

### CODE REVIEW

File: Snyk.VisualStudio.Extension.Shared/Commands/AbstractSnykCommand.cs [hunk 1]
```diff
@@ -53,6 +54,11 @@ protected AbstractSnykCommand(AsyncPackage package, OleMenuCommandService comman
         /// </summary>
         public abstract void UpdateState();
 
+        /// <summary>
+        /// Update command state asynchronously.
+        /// </summary>
+        public abstract Task UpdateStateAsync();
```
Comment (aldanchenko) [inline Snyk.VisualStudio.Extension.Shared/Commands/AbstractSnykCommand.cs:60]: It's confusing to have this two methods. Can we use only one method? Or we can't do without `UpdateState()`?

File: Snyk.VisualStudio.Extension.Shared/Commands/SnykScanCommand.cs [hunk 2]
```diff
@@ -54,8 +54,11 @@ public static async Task InitializeAsync(AsyncPackage package)
         }
 
         /// <inheritdoc/>
-        public override void UpdateState()
+        public override void UpdateState() => ThreadHelper.JoinableTaskFactory.RunAsync(this.UpdateStateAsync);
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/Commands/SnykScanCommand.cs:57]: Change to `Run`

File: Snyk.VisualStudio.Extension.Shared/Commands/AbstractSnykCommand.cs (hunk 1 above)
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/Commands/AbstractSnykCommand.cs:60]: It's common to have both non-async & async versions of the same method (like `ReadLine `and `ReadLineAsync`) so you can invoke it with and without blocking.

It's also recommended by the analyzer:
https://github.com/Microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD104.md

File: Snyk.VisualStudio.Extension.Shared/Commands/SnykStopCurrentTaskCommand.cs [hunk 3]
```diff
@@ -43,8 +43,12 @@ public static async Task InitializeAsync(AsyncPackage package)
         }
 
         /// <inheritdoc/>
-        public override void UpdateState()
+        public override void UpdateState() => ThreadHelper.JoinableTaskFactory.RunAsync(this.UpdateStateAsync);
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/Commands/SnykStopCurrentTaskCommand.cs:46]: Change to `Run`

File: Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/DownloadState.cs [hunk 4]
```diff
@@ -19,31 +20,26 @@ public override void HideComponents()
         {
             this.ToolWindowControl.HideMainMessage();
 
-            this.ToolWindowControl.Dispatcher.Invoke(() =>
+            ThreadHelper.JoinableTaskFactory.Run(async () =>
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/DownloadState.cs:23]: Switch to UI thread

File: Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/DownloadState.cs [hunk 5]
```diff
@@ -19,31 +20,26 @@ public override void HideComponents()
...
 
         /// <summary>
         /// Display progress bar and download message components.
         /// </summary>
         public override void DisplayComponents()
         {
-            this.ToolWindowControl.Dispatcher.Invoke(() =>
+            ThreadHelper.JoinableTaskFactory.Run(async () =>
             {
                 this.ToolWindowControl.progressBar.Value = 0;
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/DownloadState.cs:38]: Switch to UI thread

File: Snyk.VisualStudio.Extension.Shared/Commands/AbstractSnykCommand.cs (hunk 1 above)
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/Commands/AbstractSnykCommand.cs:56]: Removed UpdateState as it's no longer necessary

File: Snyk.VisualStudio.Extension.Shared/Commands/AbstractTaskCommand.cs [hunk 6]
```diff
@@ -21,18 +21,18 @@ public AbstractTaskCommand(AsyncPackage package, OleMenuCommandService commandSe
...
+        protected bool IsButtonAvailable()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            return Common.Guid.IsValid(SnykVSPackage.ServiceProvider.Options.ApiToken)
+                   && SnykSolutionService.Instance.IsSolutionOpen();
+        }
 
         /// <inheritdoc/>
-        protected override void OnBeforeQueryStatus(object sender, EventArgs e) => this.UpdateState();
+        protected override void OnBeforeQueryStatus(object sender, EventArgs e) => ThreadHelper.JoinableTaskFactory.Run(this.UpdateStateAsync);
```
Comment (aldanchenko) [inline Snyk.VisualStudio.Extension.Shared/Commands/AbstractTaskCommand.cs:36]: Nice ❤️

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralSettingsUserControl.cs [hunk 7]
```diff
@@ -178,6 +94,47 @@ public void Authenticate(Action<string> successCallbackAction, Action<string> er
             Logger.Information("Leave Authenticate method");
         }
 
+        private async Task OnAuthenticationSuccessfulAsync(string apiToken)
+        {
+            Logger.Information("Enter authenticate successCallback");
+
+            //TODO - try to await SwitchToMainThread
```
Comment (aldanchenko) [inline Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralSettingsUserControl.cs:101]: Don't sure did you want to do it in this PR or later?

**Review by aldanchenko — DISMISSED** [review]
LGTM 👍

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralSettingsUserControl.cs [hunk 8]
```diff
@@ -204,32 +206,84 @@ private void AuthenticateButton_Click(object sender, EventArgs eventArgs)
             this.authenticateButton.Enabled = false;
 
             Logger.Information("Start run task");
+            await TaskScheduler.Default;
```
[the remark this answers is not in the document]
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralSettingsUserControl.cs:209]: It switched **from** the UI thread to a background thread from the thread-pool. It's like `Task.Run()`

**Review by aldanchenko — APPROVED** [review]
Great work 💪

**Pull request merged**